rpcs3/Emu/Cell/lv2/sys_net/lv2_socket_p2ps.cpp

```diff
@@ -71,23 +70,35 @@ class tcp_timeout_monitor
 
 	void operator()()
 	{
+		u32 old_value = 0;
 		while (thread_ctrl::state() != thread_state::aborting)
 		{
-			std::unique_lock<std::mutex> lock(data_mutex);
-
-			if (thread_ctrl::state() == thread_state::aborting)
+			std::optional<atomic_wait_timeout> timeout;
 			{
-				break;
+				std::lock_guard lock(data_mutex);
+				if (!msgs.empty())
+				{
+					const auto current_timepoint = steady_clock::now();
+					const auto expected_timepoint = msgs.begin()->first;
+					timeout = current_timepoint > expected_timepoint ? std::nullopt : std::optional(static_cast<atomic_wait_timeout>(std::chrono::duration_cast<std::chrono::nanoseconds>(expected_timepoint - current_timepoint).count()));
+				}
+				else
+				{
+					timeout = atomic_wait_timeout::inf;
+				}
 			}
 
-			if (msgs.size())
-				wakey.wait_until(lock, msgs.begin()->first);
-			else
-				wakey.wait(lock);
+			if (timeout)
+			{
+				wakey.wait(old_value, *timeout);
+				old_value = wakey;
```

Comment:
even better would to `if (!wakey || wakey.exchange(0) == 0) wakey.wait(0, *timeout);` and notification is `wakey.release(1); wakey.notify_one();` this way it won't overflow and no need to save old value.